File: flask/skimmer/views/group.py
from flask import Blueprint, make_response, request, session
from skimmer.api import flask, group

bp = Blueprint("group", __name__)


@bp.route("/<channel_id>", methods=["GET"])
@flask.protect
def fetch_groups(user_id, channel_id):
    return [
        {"id": id, "name": name, "system": system} for (id, name, system) in group.fetch_groups(user_id, channel_id)
    ]


@bp.route("/<channel_id>", methods=["POST"])
@flask.protect
def add_group(user_id, channel_id):
    group.add_group(user_id, channel_id, request.form.get("name"))
    return "", 204


@bp.route("/<channel_id>/<id>", methods=["DELETE"])
@flask.protect
def delete_group(user_id, channel_id, id):
    group.delete_group(user_id, channel_id, id)
    return "", 204

File: react/src/api/group.js
import { apiClient } from "../config";
import { useMutation, useQueryClient } from "react-query";

export const fetchGroups = async ({ queryKey }) => {
  const [, channelId] = queryKey;
  return apiClient.get(`/group/${channelId}`).then((res) => res.data);
};

const addGroup = async ({ name, channelId }) => {
  const params = new URLSearchParams({
    name: name,
  });
  return apiClient.post(`/group/${channelId}`, params, {
    headers: { "content-type": "application/x-www-form-urlencoded" },
  });
};

export const useAddGroupMutation = (channelId, setProcessing) => {
  const queryClient = useQueryClient();
  const addMutation = useMutation(addGroup, {
    onMutate: () => {
      setProcessing(true);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: channelId });
      setProcessing(false);
    },
  });
  return addMutation;
};

const deleteGroup = async ({ channelId, id }) => {
  return apiClient.delete(`/group/${channelId}/${id}`);
};

export const useDeleteGroupMutation = (channelId, setProcessing) => {
  const queryClient = useQueryClient();
  const deleteMutation = useMutation(deleteGroup, {
    onMutate: () => {
      setProcessing(true);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: channelId });
      setProcessing(false);
    },
  });
  return deleteMutation;
};
